File: Data Structures & Algorithms/valid-sudoku/submission-0.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        List<Set<Character>> rows = new ArrayList<>();
        List<Set<Character>> columns = new ArrayList<>();
        List<Set<Character>> squares = new ArrayList<>();
        for (int i = 0; i < 9; i++) {
            rows.add(new HashSet<Character>(Arrays.asList('1','2','3','4','5','6','7','8','9')));
            columns.add(new HashSet<Character>(Arrays.asList('1','2','3','4','5','6','7','8','9')));
            squares.add(new HashSet<Character>(Arrays.asList('1','2','3','4','5','6','7','8','9')));
        }

        for (int i = 0; i < board.length; i++) {
            for (int j = 0; j < board[0].length; j++) {
                if (board[i][j] != '.') {
                    char num = board[i][j];
                    int squareNum = i / 3 * 3 + j / 3;
                    if (!rows.get(i).remove(num) || !columns.get(j).remove(num) || !squares.get(squareNum).remove(num)) {
                        return false;
                    }

                    /**
                    i / 3 * 3 + j / 3

                    0, 0 -> 0
                    2, 2 -> 0
                    3, 3 -> 4
                    */
                }
            }
        }

        return true;
    }
}

File: Data Structures & Algorithms/search-for-word/submission-0.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows = len(board)
        columns = len(board[0])

        used = set()
        def search(i: int, j: int, word: str) -> bool: 
            if not word: 
                return True

            if i < 0 or j < 0 or i >= rows or j >= columns or (i,j) in used:
                return False

            char = word[0]
            if board[i][j] != char:
                return False
            used.add((i,j))

            found = search(i+1, j, word[1:]) or search(i-1, j, word[1:]) or search(i, j + 1, word[1:]) or search(i, j-1, word[1:])

            used.remove((i,j))
            return found
        
        for i in range(rows):
            for j in range(columns):
                if search(i,j,word):
                    return True

        return False